### ISSUE TITLE
TypeTag of path-dependent type cannot be inferred by the compiler

### ISSUE THREAD

**Issue by tribbloid** [scala/bug#12113]
This problem has been recorded in:

https://stackoverflow.com/questions/59708880/in-scala-why-it-is-impossible-to-infer-typetag-from-type-alias-or-dependent-typ

A similar problem that may have the same cause has been reported in:

https://github.com/fthomas/singleton-ops/issues/152

## reproduction steps

using Scala 2.13.3 OR 2.12.11,

```scala
object TTagFromPathDependentType {

  import universe._

  def infer(): Unit = {
    type U = (Int, String)

    val ttg1 = implicitly[TypeTag[(Int, String)]]

    val ttg2 = implicitly[TypeTag[U]]
  }
}
```

## problem

Compiler throws the following error:

```
[Error] /home/peng/git/shapesafe/spike/src/main/scala/edu/umontreal/kotlingrad/spike/arity/TTagFromPathDependentType.scala:15: implicit error;
!I e (No TypeTag available for U): TypeTag[(Int,String)]
one error found

FAILURE: Build failed with an exception.
```

If the compiler is working properly it should have easily found the TypeTag for `(Int, String)`

**Comment by Jasper-M** [comment]
Related work https://github.com/scala/scala/pull/5836

**Comment by SethTisue** [comment]
are you sure the issue is path-dependence, and not that the type alias is local to a block?

```
scala 2.13.3> object O { type U = (Int, String) }
object O

scala 2.13.3> implicitly[TypeTag[O.U]]
val res4: reflect.runtime.universe.TypeTag[O.U] = TypeTag[O.U]

scala 2.13.3> { type U = (Int, String); implicitly[TypeTag[U]] }
                                                  ^
              error: No TypeTag available for U
```

**Comment by Jasper-M** [comment]
@SethTisue The problem is with "free" types, 
of which local types are an example. But also path dependent types whose definition is not statically accessible (for lack of a better specification):

```
val a = new { type T = Int }
typeOf[a.T]

trait Foo { type T }
val foo = new Foo { type T = Int }
typeOf[foo.T]
```

**Comment by tribbloid** [comment]
@Jasper-M  are you saying that once I moved those definition into a class or an object instead of a function the error will disappear?

**Comment by Jasper-M** [comment]
> are you saying that once I moved those definition into a class or an object instead of a function the error will disappear?

If you move the type for which you want a `TypeTag` out of the function it should work yes.

**Comment by SethTisue** [comment]
I'm inclined to just close this, on the grounds that it's an understandable and reasonable limitation, not a bug or shortcoming.

**Comment by tribbloid** [comment]
Yes I agree it is a free type, but it is also a type alias that can be immediately de-aliased to a dependent type.

Theoretically this is not a variable binding, so no guarantee for type equivalence, but developers may still perceive it that way
